### PR TITLE
Convert front end to Flux architecture

### PR DESCRIPTION
**Pull request by brycejacobs**

### CODE REVIEW

**Comment by popasquat89** [comment]
Ping @dariusriggins

File: app/flux/API/DashAPI.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+
+var Dash = {
+  getData: (courseName) => {
+    return fetch('/api/course/' + courseName, { method: 'GET' })
+      .then(response => {
+          return response.json();
+      })
+      .then(response => {
```
Comment (dariusriggins) [inline app/flux/API/DashAPI.js]: In this and all the others this set of thens is not necessary:

````
return fetch(...)
  .then(response => (response.json()));
````

File: app/flux/API/ExerciseAPI.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
+var ExerciseStore = require('../stores/ExerciseStore');
+
+var exercise = {
+    get: {
+        exercise: (category, total, training, tree, type) => {
+            var options = { method: 'GET' };
+            var url = '/api/exercise';
```
Comment (dariusriggins) [inline app/flux/API/ExerciseAPI.js]:
````
var url = `/api/exercise?trainingId=${training}&category=${category} and so forth`. 
````

File: app/flux/API/DashAPI.js (hunk 1 above)
Comment (popasquat89) [inline app/flux/API/DashAPI.js]: The response.json is a promise that parses the body blob.

File: app/flux/API/ExerciseAPI.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+                .then(response => {
+                    return response.json();
+                })
+                .then(response => {
+                    return Promise.resolve(response);
+                })
+                .catch(e => Promise.reject(e));
+        },
+        question: (questionId, answer) => {
+            var exercise = ExerciseStore.exercise;
```
Comment (dariusriggins) [inline app/flux/API/ExerciseAPI.js]: Feels like an anti-pattern to be performing any logic here, seems that the exercise and question should just be passed in.

File: app/flux/API/ExerciseAPI.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+                },
+                body: JSON.stringify({
+                    id: question.result.id,
+                    chosen: answer.id,
+                    exerciseId: exercise.id,
+                    questionId: question.id, 
+                })
+            };
+
+            return fetch('/api/exercise/question/' + questionId, options)
```
Comment (dariusriggins) [inline app/flux/API/ExerciseAPI.js]: ``` `/api/exercise/question/${questionId}` ```

File: app/flux/API/UserAPI.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
+var User = {
+    get: () => {
+        return fetch('/api/users')
```
Comment (dariusriggins) [inline app/flux/API/UserAPI.js]: I find it odd that all the other stores follow a Module > method (get/put/post) -> operation pattern except this one.

File: app/flux/API/UserAPI.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+            .then(response => { return response.json(); })
```
Comment (dariusriggins) [inline app/flux/API/UserAPI.js]: Can shorthand this to be ```.then(response => (response.json()))```.

File: app/flux/API/UserAPI.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+            .then(response => {
+                return Promise.resolve(response.isAuthorized ? response.user : null);
```
Comment (dariusriggins) [inline app/flux/API/UserAPI.js]: Don't need to return ```Promise.resolve```, just the user or null should suffice.

File: app/flux/API/UserAPI.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+            headers: {
+              'Accept': 'application/json',
+              'Content-Type': 'application/json'
+            },
+            body: JSON.stringify({ email: email })
+        };
+
+        return fetch('/api/signup', params)
+            .then((response) => { 
+                if (response.status >= 200 && response.status < 305){
```
Comment (dariusriggins) [inline app/flux/API/UserAPI.js]: 305? Huh? Also don't need to return a promise, just return or throw.

File: app/flux/App.js [hunk 9]
```diff
@@ -0,0 +1,71 @@
+// var API = require('./API');
```
Comment (dariusriggins) [inline app/flux/App.js]: Remove?

File: app/flux/App.js [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+var Footer = require('./components/Footer.react');
+var Header = require('./components/Header.react');
+
+var UserAPI = require('./API/UserAPI');
+var UserStore = require('./stores/UserStore');
+
+function getStateFromStores() {
+  return {
+    user: UserStore.user
```
Comment (dariusriggins) [inline app/flux/App.js]: Is this intended to be a method call? If so can you call it explicitly.

File: app/flux/App.js [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+  }
+}
+
+class App extends React.Component {
```
Comment (dariusriggins) [inline app/flux/App.js]: I'm personally really against the class syntax with React right now, there's no good way to specify mixins and the benefit from using them, especially PureMixin which you should be using here far outweighs the minor syntactic sugar it gives.

File: app/flux/App.js [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+  constructor(props){
+    super(props);
+
+    // Only lifecycle methods receive the inheritance from React.Component.
+    this._onChange = this._onChange.bind(this);
+
+    this.state = getStateFromStores();
```
Comment (dariusriggins) [inline app/flux/App.js]: Shouldn't this be done with ```getInitialState```?

File: app/flux/App.js [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  componentDidMount() {
+    UserStore.addChangeListener(this._onChange);
+    
+    if(!this.state.user){
+      // TODO(Bryce) - This should be a server action that emits event.
+      UserStore.isPending = true;
```
Comment (dariusriggins) [inline app/flux/App.js]: Flux dictates that all logic should be in the action handlers, setting pending here or calling an API and waiting for a response is an anti-pattern.

File: app/flux/App.js [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+    }
+  }
+
+  componentWillUnmount() {
+    UserStore.removeChangeListener(this._onChange);
+  }
+
+  render() {
+    return (
+      <div className="open">
```
Comment (dariusriggins) [inline app/flux/App.js]: class of open? What is that for?

File: app/flux/App.js [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+    user: UserStore.user
+  }
+}
+
+class App extends React.Component {
+  constructor(props){
+    super(props);
+
+    // Only lifecycle methods receive the inheritance from React.Component.
+    this._onChange = this._onChange.bind(this);
```
Comment (dariusriggins) [inline app/flux/App.js]: This isn't necessary if you ditch the class syntax.

File: app/flux/actions/AppActionCreators.js [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+var AppStore = require('../stores/AppStore');
+var UserAPI = require('../API/UserAPI');
+
+module.exports = {
+
+    /**
+     * The navigation menu has been clicked, 
+     *
+     */
+    clickNavigationMenu: () => {
```
Comment (dariusriggins) [inline app/flux/actions/AppActionCreators.js]: I don't understand why this is even here, this is a simple UI menu action that could be handled from within the view.

File: app/flux/actions/AppActionCreators.js [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+                type: ActionTypes.NAVIGATION_MENU_CLOSE
+            });            
+        }
+    },
+
+    /**
+     * The User has submitted a signup request.
+     *
+     */
+    clickUserSignup: (email) => {
```
Comment (dariusriggins) [inline app/flux/actions/AppActionCreators.js]: Seems like this should just be ```signup``` or ```createUser```, the event shouldn't be relevant.

File: app/flux/actions/DashActionCreators.js [hunk 18]
```diff
@@ -0,0 +1,9 @@
+
+var TrainifyAppDispatcher = require('../dispatcher/TrainifyAppDispatcher');
+var TrainifyConstants = require('../constants/TrainifyConstants');
+
+var ActionTypes = ChatConstants.ActionTypes;
```
Comment (dariusriggins) [inline app/flux/actions/DashActionCreators.js]: ??

File: app/flux/actions/ExerciseActionCreators.js [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+        // Dispatch an action that notifies of the exercise answer being selected.
+        Dispatcher.handleViewAction({
+            type: ActionTypes.EXERCISE_SELECT_ANSWER,
+            chosen: answer,
+            questionId: questionId,
+            result: result
+        });
+
+        // If this question was the last to be completed, score the exercise.
+        if(ExerciseStore.isCompleted) {
```
Comment (dariusriggins) [inline app/flux/actions/ExerciseActionCreators.js]: This makes it super confusing that this is a method and not a property.

File: app/flux/components/CategoryListItem.react.js [hunk 20]
```diff
@@ -0,0 +1,43 @@
+var { Link } = require('react-router');
```
Comment (dariusriggins) [inline app/flux/components/CategoryListItem.react.js]: Seems really weird to mix AMD with imports syntax, if you're going to use it go all the way:
````
import {Link} from 'react-router';
````

File: app/flux/components/Checkout.react.js [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+var UserStore = require('../stores/UserStore');
+
+class Checkout extends React.Component{
+    constructor (props){
+        super(props);
+
+        // Only lifecycle methods receive the inheritance from React.Component.
+        this._onChange = this._onChange.bind(this);
+        
+        this.state = {
```
Comment (dariusriggins) [inline app/flux/components/Checkout.react.js]: Ditto on class syntax and getInitialState.

File: app/flux/components/Checkout.react.js [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+            training: null,
+            user: UserStore.user        
+        };
+    }
+
+    componentDidMount (){
+        UserStore.addChangeListener(this._onChange);
+
+        if (!this.state.user && UserStore.isPending){
+            setTimeout(() => {
```
Comment (dariusriggins) [inline app/flux/components/Checkout.react.js]: This seems really hacky and brittle, this should be handling dispatched events and not using setTimeout.

File: app/flux/components/Checkout.react.js [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+                if (!this.state.user){
+                    this.context.router.transitionTo('signup');    
+                }
+            }, 800).bind(this);
+        } else if (!this.state.user){
+            this.context.router.transitionTo('signup');
+        }
+
+        var trainingId = this.context.router.getCurrentParams().id;
```
Comment (dariusriggins) [inline app/flux/components/Checkout.react.js]: Just a personal preference, but I prefer to pass down the params as ```...props``` and then just use ```this.props.params.id```.

File: app/flux/components/Checkout.react.js [hunk 24]
```diff
@@ -0,0 +1,96 @@
...
+    render (){
+        var { stripe, training } = this.state;
+
+        if(!training) {
+            return <div></div>;
+        }
+
+        return (
+            <main className="checkout">
+                <form action={'/api/buy/' + training.id} method="post">
```
Comment (dariusriggins) [inline app/flux/components/Checkout.react.js]: ```action="/api/buy/{training.id}"```?

File: app/flux/components/Checkout.react.js [hunk 25]
```diff
@@ -0,0 +1,96 @@
...
+                    { training.cost ?
```
Comment (dariusriggins) [inline app/flux/components/Checkout.react.js]: This should be changed and is highly confusing, just create a function that handles each case to make this more clear.

File: app/flux/components/Dash.react.js [hunk 26]
```diff
@@ -0,0 +1,138 @@
...
+    });
+  }
+
+  componentWillUnmount() {
+    DashStore.removeChangeListener(this._onChange);
+  }
+
+  render() {
+    if(this.state.isEmpty){
+      return (<div style={{'minHeight': 500 + 'px'}}></div>);
```
Comment (dariusriggins) [inline app/flux/components/Dash.react.js]: Inline style really necessary?

File: app/flux/components/Dash.react.js [hunk 27]
```diff
@@ -0,0 +1,138 @@
...
+    var trees = stats.trees;
+    var user = this.state.user;
+    
+    // Create our List Items that will present the categories.
+    var categoryList = categories.map(category => {
+      return (<CategoryListItem
+                key = {category.id}
+                category = {category}
+                trainingId = {training.id} />);      
+    }, this);
```
Comment (dariusriggins) [inline app/flux/components/Dash.react.js]: ``, this)`` is not necessary, fat arrow functions always bind to this.

File: app/flux/components/Exercise.react.js [hunk 28]
```diff
@@ -0,0 +1,197 @@
...
+                      fill="none"
+                      strokeDasharray="282, 282"
+                      strokeWidth="8"></path>
+                    <text id='progress-percent' x="9" y="21">
+                    {progress.toFixed() + '%'}
+                    </text>
+                 </svg>
+            </div>
+            <div className="exercise-overview">
+                <a href="/">Go back to dashboard</a>
```
Comment (dariusriggins) [inline app/flux/components/Exercise.react.js]: Should use ```<Link>```.

File: app/flux/components/Exercise.react.js [hunk 29]
```diff
@@ -0,0 +1,197 @@
...
+            </div>
+          </div>
+        </Headroom>
+
+        <main className="exercise">
+            <div className={cx({review: isReview},{hidden: !isReview})}>
```
Comment (dariusriggins) [inline app/flux/components/Exercise.react.js]: We need a package to figure out which class to use?

File: app/flux/components/Exercise.react.js [hunk 30]
```diff
@@ -0,0 +1,197 @@
...
+            </ul>
+
+            <p className="legal">{training.legal}</p>
+        </main>
+      </div>
+    );
+  }
+
+  _locationReload() {
+    location.reload();
```
Comment (dariusriggins) [inline app/flux/components/Exercise.react.js]: What's the purpose of this?

File: app/flux/components/Header.react.js [hunk 31]
```diff
@@ -0,0 +1,117 @@
...
+    return (
+      <div>
+        <header className={cx('app-bar', 'promote-layer', {open: navDrawerOpen})}>
+          <div className="app-bar-container">
+            <button className="menu" onClick={this._onMenuClick}>
+              <img src="/images/hamburger.svg" width="24" height="24" alt="Menu"/>
+            </button>
+            <h1 className="logo"></h1>
+            <section className="app-bar-actions">
+              <a href={loginLink} className={cx('app-bar-user', {'signed-in': loggedIn})}>
```
Comment (dariusriggins) [inline app/flux/components/Header.react.js]: Not sure why we're not using ```<Link>``` here.

File: app/flux/components/Header.react.js [hunk 32]
```diff
@@ -0,0 +1,117 @@
...
+            <h1 className="logo">
+              <a href="/"><img src="/images/logo.svg" alt="Trainify" width="166" height="77" /></a>
+            </h1>
+          </div>
+        </header>
+
+        <nav className={cx('navdrawer-container', 'promote-layer', {open: navDrawerOpen})}>
+          <h4>Navigation</h4>
+          <ul>
+            <li><a href="/">Home</a></li>
```
Comment (dariusriggins) [inline app/flux/components/Header.react.js]: Ditto on ```<Link>```.

File: app/flux/API/DashAPI.js (hunk 1 above)
Comment (dariusriggins) [inline app/flux/API/DashAPI.js]: I meant the ones after it, you only need the code block I put in the comment, not the Promise.resolve and catch.

**Comment by dariusriggins** [comment]
There seems to be a ton of boilerplate and one off patterns that are going to make this difficult to maintain and understand. It seems like adopting a library that implements flux would cut down on that significantly. Based on my research I'd suggest alt or flummox. I think flummox is where I'm personally leaning, but w/e, your choice. If you just want to fix this up and land it that's fine too.

File: app/flux/components/Exercise.react.js (hunk 30 above)
Comment (popasquat89) [inline app/flux/components/Exercise.react.js]: Retake the same exercise.

File: app/flux/components/CategoryListItem.react.js (hunk 20 above)
Comment (popasquat89) [inline app/flux/components/CategoryListItem.react.js]: Module syntax was benched for ES6, this is just simply getting the prototype class Link from the module.

File: app/flux/components/Checkout.react.js (hunk 22 above)
Comment (popasquat89) [inline app/flux/components/Checkout.react.js]: Yeah, not a fan of this. I'm going to do some fixup to this.

File: app/flux/components/Exercise.react.js (hunk 29 above)
Comment (popasquat89) [inline app/flux/components/Exercise.react.js]: Could use string concat, but this makes it a bit cleaner imo.

File: app/flux/components/Header.react.js (hunk 31 above)
Comment (popasquat89) [inline app/flux/components/Header.react.js]: I was having a problem passing in classes to Link

**Pull request merged**